tests/routes/users.test.js

```diff
@@ -0,0 +1,184 @@
+        expect(response.statusCode).toBe(201);
+        expect(response.body.name).toBe(data.name);
+
+        // clean up DB
+        await User.findOneAndDelete({}, {
+            sort: {
+                createdAt: -1
+            }
+        })
+        done();
+    })
+
+    it('Should return 403 when dob is invalid', async done => {
+        const data = {
+            name: 'John Doe',
+            dob: '1900-01-1000',
+            address: 'Up in heaven',
+            description: 'I do not exist'
+        };
+        const response = await request
+            .post('/users')
+            .send(data);
+        expect(response.statusCode).toBe(403);
+        expect(response.body.message).toBe('Invalid parameters.');
+
+        done();
+    })
+
+    it('Should return 403 when required fields are not fulfilled', async done => {
+        const data = {
+            name: 'John Doe',
+            dob: '1900-01-10',
+            description: 'I do not exist'
+        };
+        const response = await request
+            .post('/users')
+            .send(data);
+        expect(response.statusCode).toBe(403);
+        expect(response.body.message).toBe('Invalid parameters.');
+
+        done();
+    })
+
+    it('Should find the user by ID', async done => {
+        const response = await request
+            .get('/users/6088d0ccf5a4bb73ed24cf3f');
+
+        expect(response.statusCode).toBe(200);
+        expect(response.body.name).toBe('Marcelo Saciloto');
+
+        done();
+    })
+
+    it('Should return 404 when user does not exist', async done => {
+        const response = await request
+            .get('/users/6088d0ccf5a4bb73ed24cfee');
+
+        expect(response.statusCode).toBe(404);
+        expect(response.body.message).toBe('No valid entry found for provided ID.');
+
+        done();
+    })
+
+    it('Should update user by ID', async done => {
+        const data = {
+            name: 'Juan Mendosa',
+            description: 'I am the king of narcos.'
+        }
+
+        const response = await request
+            .patch('/users/6088d0ccf5a4bb73ed24cf3f')
+            .send(data);
+
+        expect(response.statusCode).toBe(200);
+        expect(response.body.name).toBe(data.name);
+        expect(response.body.description).toBe(data.description);
+
+        done();
+    })
+
+    it('Should delete user by ID', async done => {
+        const response = await request
+            .delete('/users/6088d0ccf5a4bb73ed24cf1f');
+
+        expect(response.statusCode).toBe(204);
+
+        done();
+    })
+
+    it('Should return 404 when trying to delete a nonexistent user', async done => {
+        const response = await request
+            .delete('/users/6088d0ccf5a4bb73ed24cfee');
+
+        expect(response.statusCode).toBe(404);
+        expect(response.body.message).toBe('User not found.');
+
+        done();
+    })
+
+    it('Should return 500 when trying to delete user by invalid ID', async done => {
```

Comment:
can you explain why 500 status code is chosen in this case?
some people also suggest status code like 404, 200, 204 when ID is invalid, can you compare the pros and cons of these status codes?

Reply:
I think it's just a typo here. In my opinion using 404 makes more sense because it indicates the failure of operation explicitly, however 200 and 204 are commonly used then the operation succeeds.